scrapypuppeteer/middleware.py

```diff
@@ -121,20 +136,23 @@ def _serialize_body(self, action, request):
             return json.dumps(payload)
         return str(payload)
 
-    def process_response(self, request, response, spider):
+    def process_response(self, request: Request, response, spider):
         if not isinstance(response, TextResponse):
             return response
 
         puppeteer_request = request.meta.get('puppeteer_request')
         if puppeteer_request is None:
             return response
 
+        request = request.replace(url=puppeteer_request.url,
+                                  method=puppeteer_request.action.endpoint.upper())
```

Comment:
Puppeteer action is not a [standard HTTP method](https://www.rfc-editor.org/rfc/rfc9110.html#section-9.1-7). While it is not likely that someone tries to repeat it, doing so will result in unexpected behavior. I think it would be better to implement custom Request subclass with overridden repr.

Ultimately I believe we should try replacing this request with original PuppeteerRequest.

Reply:
Replacing request with original PuppeteerRequest is not fully solves problems. In this case we still will get <POST localhost: ...> in case we fail our request. For example, when RetryMiddleware works it uses not failed response but its request, so we get <POST localhost ...>.
We definitely need interim CustomRequest class for this and send it in ServiceMiddleware in process_request